clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java

```diff
@@ -127,6 +128,24 @@ public void awaitMetadataUpdate() {
     }
 
     /**
+     * Block until the metadata has been refreshed or the timeout has expired.
+     * @param timeout The maximum duration (in ms) to wait for the request
+     * @throws TimeoutException if the timeout has expired
+     */
+    public void awaitMetadataUpdate(long timeout) {
+        long now = time.milliseconds();
+        long deadline = now + timeout;
+        int version = this.metadata.requestUpdate();
+        while (this.metadata.version() == version && now < deadline) {
```

Comment:
One question: what is our plan for cases where the metadata update fails? A few options that I can think of:
1. Attempt it again until the timeout expires
2. Throw an exception immediately
3. Wait until the timeout expires and then throw an exception

If I understand correctly, we are doing 3 at the moment (please correct me if I am wrong). I think 1 or 2 would be better although perhaps more work. We can talk about implementation issues after we decide what we want to do though. cc @hachikuji